Services/Database/classes/Atom/class.ilAtomQuery.php

```diff
@@ -0,0 +1,320 @@
+	 */
+	protected static $anomalies_map = array(
+		self::ISOLATION_READ_UNCOMMITED => array(
+			self::ANO_LOST_UPDATES,
+			self::ANO_DIRTY_READ,
+			self::ANO_NON_REPEATED_READ,
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_READ_COMMITED   => array(
+			self::ANO_NON_REPEATED_READ,
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_REPEATED_READ   => array(
+			self::ANO_PHANTOM,
+		),
+		self::ISOLATION_SERIALIZABLE    => array(),
+	);
+	/**
+	 * @var int
+	 */
+	protected $isolation_level = self::ISOLATION_SERIALIZABLE;
+	/**
+	 * @var array
+	 */
+	protected $tables = array();
+	/**
+	 * @var callable
+	 */
+	protected $query;
+	/**
+	 * @var Closure[]
+	 */
+	protected $queries = array();
+	/**
+	 * @var
+	 */
+	protected $ilDBInstance;
+	/**
+	 * @var ilAtomQuery
+	 */
+	protected static $instance;
+
+
+	/**
+	 * ilAtomQuery constructor.
+	 *
+	 * @param \ilDBInterface $ilDBInstance
+	 * @param int $isolation_level currently only ISOLATION_SERIALIZABLE is available
+	 */
+	public function __construct(ilDBInterface $ilDBInstance, $isolation_level = self::ISOLATION_SERIALIZABLE) {
+		$this->ilDBInstance = $ilDBInstance;
+		$this->isolation_level = $isolation_level;
+	}
+
+	//
+	//
+	//
+	/**
+	 * @return array
+	 */
+	public function getRisks() {
+		return static::getPossibleAnomalies($this->getIsolationLevel());
+	}
+
+
+	/**
+	 * Add table-names which are influenced by your queries, MyISAm has to lock those tables. Lock
+	 *
+	 * @param string $table_name
+	 * @param int $lock_level use ilAtomQuery::LOCK_READ or ilAtomQuery::LOCK_WRITE
+	 * @throws \ilDatabaseException
+	 */
+	public function addTable($table_name, $lock_level) {
+		if (!in_array($lock_level, array( static::LOCK_READ, static::LOCK_WRITE ))) {
+			throw new ilDatabaseException('The current Isolation-level does not support the desired lock-level. use ilAtomQuery::LOCK_READ or ilAtomQuery::LOCK_WRITE');
+		}
+		$this->tables[] = array( $table_name, $lock_level );
+	}
+
+
+	/**
+	 * Every action on the database during this isolation has to be passed as Closure to ilAtomQuery.
+	 * An example:
+	 * $ilAtomQuery->addQueryClosure( function ($ilDB) use ($new_obj_id, $current_id) {
+	 *        $ilDB->manipulateF('
+	 *            DELETE FROM frm_user_read
+	 *            WHERE obj_id = %s AND thread_id =%s',
+	 *        array('integer', 'integer'),
+	 *        array($new_obj_id, $current_id));
+	 *    });
+	 *
+	 * @param \Closure $query
+	 */
+	public function addQueryClosure(Closure $query) {
+		$this->queries[] = $query;
+	}
+
+
+	/**
+	 * Fire your Queries
+	 *
+	 * @throws \ilDatabaseException
+	 */
+	public function run() {
+		self::checkIsolationLevel($this->getIsolationLevel());
+		$this->checkQueries();
+
+		if ($this->hasWriteLocks() && $this->getIsolationLevel() != self::ISOLATION_SERIALIZABLE) {
+			throw new ilDatabaseException('The selected Isolation-level is not allowd when locking tables with write-locks');
+		}
+
+		if ($this->ilDBInstance->supportsTransactions()) {
```

Comment:
The current version is based on ilDBInterface::runAtomQuery(); The points I  don't like in this version is the fact that you have to know the structure of the array for the locks (as in the old implementation) and you can't seperate queries to multiple callables which would allow to run them only once when one is failing.

Both of those points could be optimized by provide method chaining...